### PR TITLE
Configvlan with nmcli

### PR DESCRIPTION
**Pull request by zet809**
### The content of the PR:
* [x] implement subroutine create_vlan_interface_nmcli

### The UT result
1. On Peer Node:
```
# ip a show dev ens6.194
111: ens6.194@ens6: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 42:2d:0a:04:1c:67 brd ff:ff:ff:ff:ff:ff
    inet 194.4.28.103/8 brd 194.255.255.255 scope global noprefixroute ens6.194
       valid_lft forever preferred_lft forever
    inet6 fe80::f1a0:df4f:66ef:a65d/64 scope link noprefixroute
       valid_lft forever preferred_lft forever
```
2. For CN to config:
```
# lsdef c910f04x28v04 | grep nic
    nicdevices.ens6.194=ens6
    nichostnamesuffixes.ens6.194=-ens6-194
    nicips.ens6.194=194.4.28.104
    nicnetworks.ens6.194=net194
    nictypes.ens6=ethernet
    nictypes.ens6.194=vlan
```
3. updatenode to configure vlan:
```
# updatenode c910f04x28v04 -P confignetwork
c910f04x28v04: =============updatenode starting====================
c910f04x28v04: trying to download postscripts...
c910f04x28v04: postscripts downloaded successfully
c910f04x28v04: trying to get mypostscript from 10.4.28.101...
c910f04x28v04: postscript start..: confignetwork
c910f04x28v04: [I]: NetworkManager is active
c910f04x28v04: [I]:
c910f04x28v04: [I]:
c910f04x28v04: [I]: All valid nics and device list:
c910f04x28v04: [I]: ens6.194 ens6
c910f04x28v04: ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
c910f04x28v04: configure nic and its device : ens6.194 ens6
c910f04x28v04: [I]: create_vlan_interface_nmcli ifname=ens6 vlanid=194 ipaddrs=194.4.28.104
c910f04x28v04: [I]: Pickup xcatnet, "net194", from NICNETWORKS for interface "ens6".
c910f04x28v04: Connection 'xcat.ens6.194' (fed2c740-bee4-4c07-9694-87f26cb13877) successfully added.
c910f04x28v04: [I]: nmcli con add type vlan con-name xcat.ens6.194 dev ens6 id 194 method none ip4 194.4.28.104/8
c910f04x28v04: Connection successfully activated (D-Bus active path: /org/freedesktop/NetworkManager/ActiveConnection/2525)
c910f04x28v04: postscript end....: confignetwork exited with code 0
c910f04x28v04: Running of postscripts has completed.
c910f04x28v04: =============updatenode ending====================
```
4. On target CN to ping peer node:
Before configure
```
# ip a |grep ens6
5: ens6: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
```
After configure
```
# ip a |grep ens6
5: ens6: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
18: ens6.194@ens6: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    inet 194.4.28.104/8 brd 194.255.255.255 scope global noprefixroute ens6.194
# ping 194.4.28.103 -c 2
PING 194.4.28.103 (194.4.28.103) 56(84) bytes of data.
64 bytes from 194.4.28.103: icmp_seq=1 ttl=64 time=0.265 ms
64 bytes from 194.4.28.103: icmp_seq=2 ttl=64 time=0.339 ms

--- 194.4.28.103 ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 55ms
rtt min/avg/max/mdev = 0.265/0.302/0.339/0.037 ms
```

### CODE REVIEW

File: xCAT/postscripts/nicutils.sh [hunk 1]
```diff
@@ -1623,7 +1611,86 @@ function check_NetworkManager_or_network_service() {
...
+    if nmcli con |grep "^$con_name " > /dev/null; then
+        tmp_con_name=$con_name."-tmp"
+        nmcli con modify $con_name connection.id $tmp_con_name
+    fi
+    cmd="nmcli con add type vlan con-name $con_name dev $ifname id $(( 10#$vlanid )) $_ipaddrs $_mtu"
+    $cmd
+    log_info "$cmd"
+    nmcli con up $con_name
+
+    i=0
```
Comment (bybai) [inline xCAT/postscripts/nicutils.sh]: from line 1673 to line 1692 can be abstract as a common function, this function is common for bridge and bond.

**Comment by bybai** [comment]
Hi @zet809,  I created PR https://github.com/xcat2/xcat-core/pull/6106 including the following common functions, create_vlan_...... should also including corresponding check.
1. check if vlan base network device managed or not , if not managed, use nmcli set it be managed
2. check if the first ipv4 ip addr for vlan device is valid or not from nics.ips

File: xCAT/postscripts/confignetwork [hunk 2]
```diff
@@ -574,10 +574,11 @@ function configure_nicdevice {
                 vlanid=`echo "$nic_dev" | $sed -e 's/^\(.*\)vla\?n\?\([0-9]\+\)$/\2/'`
                 vlanname=`echo "$nic_dev" | $sed -e 's/^\(.*\)vla\?n\?\([0-9]\+\)$/\1/'`
             fi
+            ipaddrs=$(find_nic_ips $nic_dev)
             if [ "$networkmanager_active" = "0" ]; then
-                create_vlan_interface ifname=$vlanname vlanid=$vlanid
+                create_vlan_interface ifname=$vlanname vlanid=$vlanid ipaddrs=$ipaddrs
```
Comment (bybai) [inline xCAT/postscripts/confignetwork]: ipaddrs=$ipaddrs is not needed in old create_vlan_interface.

File: xCAT/postscripts/nicutils.sh [hunk 3]
```diff
@@ -1606,12 +1605,92 @@ function check_NetworkManager_or_network_service() {
 #
 # create vlan using nmcli
 #
-# input : ifname=<ifname> slave_ports=<ports> xcatnet=<xcatnetwork> _ipaddr=<ip> _netmask=<netmask> _mtu=<mtu> _bridge=<bridge_name> vlanid=<vlanid>
+# input : ifname=<ifname> vlanid=<vlanid>
```
Comment (bybai) [inline xCAT/postscripts/nicutils.sh]: `input : ifname=<ifname> vlanid=<vlanid>` should be `input : ifname=<ifname> vlanid=<vlanid> ipaddrs=<ipaddrs>`

File: xCAT/postscripts/confignetwork [hunk 4]
```diff
@@ -574,10 +574,11 @@ function configure_nicdevice {
...
             elif [ "$networkmanager_active" = "1" ]; then
-                create_vlan_interface_nmcli ifname=$vlanname vlanid=$vlanid
+                create_vlan_interface_nmcli ifname=$vlanname vlanid=$vlanid ipaddrs=$ipaddrs
```
Comment (bybai) [inline xCAT/postscripts/confignetwork]: `create_vlan_interface_nmcli` return code should be checked here. If create vlan failed, `errorcode=1` should be set here, `errorcode` is global variable to control `confignetwork` return code.  I think old function also miss this check, it is an old defect.

File: xCAT/postscripts/nicutils.sh [hunk 5]
```diff
@@ -1606,12 +1605,92 @@ function check_NetworkManager_or_network_service() {
...
+    fi
+
+    if [ ! -z "$ipaddrs" ]; then
+        _netmask_long=`get_network_attr $_xcatnet mask`
+        if [ $? -ne 0 ]; then
+            log_error "No valid netmask get for $ifname.$vlanid"
+            return 1
+        else
+            _netmask=$(v4mask2prefix $_netmask_long)
+            _ipaddrs="method none ip4 $ipaddrs/$_netmask"
```
Comment (bybai) [inline xCAT/postscripts/nicutils.sh]: Before this line, should check if $ipaddrs is the first valid ipv4 address.

File: xCAT/postscripts/nicutils.sh [hunk 6]
```diff
@@ -1606,12 +1605,92 @@ function check_NetworkManager_or_network_service() {
...
+    fi
+    #load the 8021q module if not loaded.
+    load_kmod module=8021q retry=10 interval=0.5
+    con_name="xcat.$ifname.$vlanid"
+    tmp_con_name=""
+    if nmcli con |grep "^$con_name " > /dev/null; then
+        tmp_con_name=$con_name."-tmp"
+        nmcli con modify $con_name connection.id $tmp_con_name
+    fi
+    cmd="nmcli con add type vlan con-name $con_name dev $ifname id $(( 10#$vlanid )) $_ipaddrs $_mtu"
```
Comment (bybai) [inline xCAT/postscripts/nicutils.sh]: before this line, should check if $ifname is managed or not using nmcli.

File: xCAT/postscripts/confignetwork (hunk 2 above)
Comment (zet809) [reply xCAT/postscripts/confignetwork]: OK

**Comment by zet809** [comment]
Need to take a look at why a rebase master involve in so many commits.

**Comment by zet809** [comment]
Replace this PR with https://github.com/xcat2/xcat-core/pull/6110, so close this one.

**Pull request closed without merging**